crates/bevy_ecs/src/query/state.rs

```diff
@@ -1276,43 +1339,156 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
             >())),
         }
     }
+
+    /// Gets the reasons an [`Entity`] might not match this query, see [`QueryEntityMismatchDetail`]
+    pub fn get_mismatches_detail(
+        &self,
+        components: &Components,
+        entity_location: EntityLocation,
+        archetype: &Archetype,
+    ) -> Vec<QueryEntityComponentMismatch> {
+        let mut mismatch_details = Vec::new();
+
+        for components_access in &self.component_access.filter_sets {
+            let query_with_components = components_access.with.clone();
+            let query_without_components = components_access.without.clone();
+            let mut query_unmatched_components = query_with_components.clone();
+
+            let mut query_without_components_in_entity = Vec::new();
+            for component_id in archetype.components() {
+                if query_without_components.contains(component_id.index()) {
+                    query_without_components_in_entity.push((
+                        components.get_info(component_id).unwrap().name().to_owned(),
+                        component_id,
+                    ));
+                } else if query_with_components.contains(component_id.index()) {
+                    query_unmatched_components.set(component_id.index(), false);
+                }
+            }
+
+            let query_with_components_not_in_entity = query_unmatched_components
+                .ones()
+                .map(|component_id| {
+                    let component_id = ComponentId::new(component_id);
+                    (
+                        components.get_info(component_id).unwrap().name().to_owned(),
+                        component_id,
+                    )
+                })
+                .collect();
+
+            mismatch_details.push(QueryEntityComponentMismatch {
+                query_with_components_not_in_entity,
+                query_without_components_in_entity,
+                entity_archetype: entity_location.archetype_id,
+                query_with_components,
+                query_without_components,
+            });
+        }
+
+        mismatch_details
+    }
 }
 
 /// An error that occurs when retrieving a specific [`Entity`]'s query result from [`Query`](crate::system::Query) or [`QueryState`].
-// TODO: return the type_name as part of this error
-#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[derive(Debug, PartialEq, Eq, Clone)]
 pub enum QueryEntityError {
-    /// The given [`Entity`]'s components do not match the query.
-    ///
-    /// Either it does not have a requested component, or it has a component which the query filters out.
-    QueryDoesNotMatch(Entity),
+    /// The given [`Entity`]'s components do not match the query, see [`QueryEntityMismatchDetail`].
+    QueryDoesNotMatch(QueryEntityMismatchDetail, QueryEntityErrorDetail),
```

Comment:
I think `QueryEntityMismatchDetail` is too heavy, both to compute and memory-wise to add to the return result of `Query::get`. Since `Query::get_mismatches_detail` exists, I think there is no need to add this to `QueryEntityError`, users can always call `get_mismatches_detail` if they have an `Err` and need to know the exact problems encountered.

It's very common for bevy code to do

```rust
if let Ok(foo) = query.get(entity) {
    // ...
}
```

If the `Err` is a common case, and is always skipped here, the user is going to create and discard a `QueryEntityMismatchDetail` multiple times per frame, while they absolutely do not care about it.

Reply:
I kinda hesitated a bit about this yeah.

About removing the `QueryEntityMismatchDetail` in the `QueryDoesNotMatch` variant (or the struct holding it) and letting people use `get_mismatches_detail`:
- It would be a tiny bit more annoying to use (you have to match the variant to check that it's a `QueryDoesNotMatch` before calling the function, but still need the query itself to call the function so you can't just pass the error to a helper function handle it you need to pass the query too), but I agree that it's still better than what we have now (which is no information at all about the mismatch) and that it's a usecase too specific to calculate it and store it every time.
- I would like to split the variant `QueryDoesNotMatch` into `QueryComponentsMismatch` and `QueryChangeDetectionMismatch` at least, otherwise in the case of a change detection mismatch, people might use `get_mismatches_detail` and not understand why it returns empty arrays (or `None`, see bellow).
- I need to change `get_mismatches_detail` as it's not super user friendly right now, should I make it so it returns an `Option` so it returns `None` if there's no mismatch? This would save computing time when it's the case and make it clearer that the entity does in fact match the query (at least components-wise), but maybe people would still want information about "subqueries" in the case of an `Or`?